src/webgl/p5.RendererGL.js

```diff
@@ -891,22 +887,169 @@ p5.RendererGL.prototype.loadPixels = function() {
+  fbo.updatePixels();
+  this._pInst.push();
+  this._pInst.resetMatrix();
+  this._pInst.clear();
+  this._pInst.imageMode(constants.CENTER);
+  this._pInst.image(fbo, 0, 0);
+  this._pInst.pop();
+  this.GL.clearDepth(1);
+  this.GL.clear(this.GL.DEPTH_BUFFER_BIT);
+};
+
+/**
+ * @private
+ * @returns {p5.Framebuffer} A p5.Framebuffer set to match the size and settings
+ * of the renderer's canvas. It will be created if it does not yet exist, and
+ * reused if it does.
+ */
+p5.RendererGL.prototype._getTempFramebuffer = function() {
+  if (!this._tempFramebuffer) {
+    this._tempFramebuffer = this._pInst.createFramebuffer({
+      format: constants.UNSIGNED_BYTE,
+      useDepth: this._pInst._glAttributes.depth,
+      depthFormat: constants.UNSIGNED_INT,
+      antialias: this._pInst._glAttributes.antialias
+    });
   }
+  return this._tempFramebuffer;
+};
 
-  const pd = this._pInst._pixelDensity;
-  this.GL.readPixels(
-    0, 0, this.width * pd, this.height * pd,
-    this.GL.RGBA, this.GL.UNSIGNED_BYTE,
+/**
+ * @private
+ * @param {Uint8Array|Float32Array|undefined} pixels An existing pixels array to reuse if the size is the same
+ * @param {WebGLRenderingContext} gl The WebGL context
+ * @param {WebGLFramebuffer|null} framebuffer The Framebuffer to read
+ * @param {Number} x The x coordiante to read (factoring in pixel density)
+ * @param {Number} y The y coordiante to read (factoring in pixel density)
+ * @param {Number} width The width in pixels to be read (factoring in pixel density)
+ * @param {Number} height The height in pixels to be read (factoring in pixel density)
+ * @param {GLEnum} format Either RGB or RGBA depending on how many channels to read
+ * @param {GLEnum} type The datatype of each channel, e.g. UNSIGNED_BYTE or FLOAT
+ * @param {Number|undefined} flipY If provided, the total height with which to flip the y axis about
+ * @returns {Uint8Array|Float32Array} pixels A pixels array with the current state of the
+ * WebGL context read into it
+ */
+export function readPixelsWebGL(
+  pixels,
+  gl,
+  framebuffer,
+  x,
+  y,
+  width,
+  height,
+  format,
+  type,
+  flipY
+) {
+  // Record the currently bound framebuffer so we can go back to it after, and
+  // bind the framebuffer we want to read from
+  const prevFramebuffer = gl.getParameter(gl.FRAMEBUFFER_BINDING);
+  gl.bindFramebuffer(gl.FRAMEBUFFER, framebuffer);
+
+  const channels = format === gl.RGBA ? 4 : 3;
+
+  // Make a pixels buffer if it doesn't already exist
+  const len = width * height * channels;
+  const TypedArrayClass = type === gl.UNSIGNED_BYTE ? Uint8Array : Float32Array;
+  if (!(pixels instanceof TypedArrayClass) || pixels.length !== len) {
+    pixels = new TypedArrayClass(len);
+  }
+
+  gl.readPixels(
+    x,
+    flipY ? (flipY - y - height) : y,
+    width,
+    height,
+    format,
+    type,
     pixels
   );
-};
+
+  // Re-bind whatever was previously bound
+  gl.bindFramebuffer(gl.FRAMEBUFFER, prevFramebuffer);
+
+  if (flipY) {
```

Comment:
This seems like it could be a perf bottleneck. Would we save cpu by creating a temporary fbo and just doing a the y flip in a shader, then reading it back?

Reply:
Good question, that might be faster! I'll do some tests and get back to you.

Reply:
I did a little test, and interestingly, it seems like doing it on the CPU is faster for me. Here's a test sketch which doesn't even get to the step yet of flipping the pixels, it just copies from the main canvas to the framebuffer: https://editor.p5js.org/davepagurek/sketches/J6oNWOpEz

On my M1 macbook pro I get these times:
```
CPU: 1874ms 
GPU: 3242ms 
```

So maybe we'll keep it as is for now, and in the future if we back the main canvas with a framebuffer all the time, we can just read its pixels directly instead of having to do any copying or flipping.

Reply:
Hmm, I think maybe we got some crossed wires. I was thinking that it could be faster to do the flipping just for fbos, not the backing canvas. But maybe this is all moot since (I think) we only need to do the flip when reading pixels directly from the canvas?

So the tests would be

CPU:
1. Read FBO to pixels
2. Cpu flip pixels

GPU: 
1. render FBO using a temp fbo on a quad with a flipping shader
2. Read FBO to pixels.

I tried to put this together as well but it seems like CPU still comes out on top.  
https://editor.p5js.org/aferriss/sketches/Q1AFNF1YM

Also, strangely, I noticed in your test that GPU actually came out on top when the iterations were <5. Maybe just hammering the GPU causes some throttling but when the work is less it's able to execute quickly?

Reply:
> But maybe this is all moot since (I think) we only need to do the flip when reading pixels directly from the canvas?

Ah yeah I think that's why I assumed it's the other way around. To make the main canvas appear right-side-up, it needs to store its pixels in the way it's currently doing it, which is unfortunately the opposite of 2D canvas pixels for some reason.

> I tried to put this together as well but it seems like CPU still comes out on top.

Well that's good to know! I guess although the GPU is fast, copying all the pixels from one buffer to another on the GPU is still a bit slower than rearranging data in one buffer on the CPU? So I guess we should leave the current `flipY` implementation as-is for now.